### PR TITLE
Creating a MultiApp step for the tutorial

### PR DESCRIPTION
**Pull request by aeslaughter**
(refs #5307)

### CODE REVIEW

File: framework/src/bcs/NeumannBC.C [hunk 1]
```diff
@@ -18,13 +18,13 @@ template<>
 InputParameters validParams<NeumannBC>()
 {
   InputParameters params = validParams<IntegratedBC>();
-  params.addParam<Real>("value", 0.0, "The value of the gradient on the boundary.");
+  params.addParam<PostprocessorName>("value", 0.0, "The value of the gradient on the boundary.");
```
Comment (friedmud) [inline framework/src/bcs/NeumannBC.C]: I don't like this for one of our most basic classes in MOOSE.  This class is used (and inherited from) by EVERYONE... and now it's going to show that the `value` is expecting a `PostprocessorName`?  And now Peacock is going to suggest the value be an Postprocessor in your input file?

The right thing to do here is to make a custom BC that does this in your example.

File: framework/src/base/FEProblem.C [hunk 2]
```diff
@@ -416,7 +416,7 @@ void FEProblem::initialSetup()
     Moose::setup_perf_log.push("initial adaptivity", "Setup");
     unsigned int n = adaptivity().getInitialSteps();
 
-    if (n && !_app.isUltimateMaster())
+    if (n && !_app.isUltimateMaster() && _app.isRestarting())
```
Comment (friedmud) [inline framework/src/base/FEProblem.C]: Is this a bugfix?  This should probably be a separate PR.

BTW: I might be removing this restriction altogether with the MultiApp restart work I currently have going...

**Comment by friedmud** [comment]
Can you describe, at a high level, what this MultiApp setup is supposed to be doing?

**Comment by friedmud** [comment]
BTW: It's awesome to get a tutorial step for MultiApps!  Thanks for working on this!  I know that some of my comments are negative... but I am really excited to get this added to the tutorial!

File: framework/src/base/FEProblem.C (hunk 2 above)
Comment (aeslaughter) [inline framework/src/base/FEProblem.C]: Yes this fixes a bug fix, the MultiApps that have sub-apps with initial adaptivity will not run because this error is thrown.

File: framework/src/bcs/NeumannBC.C (hunk 1 above)
Comment (aeslaughter) [inline framework/src/bcs/NeumannBC.C]: I will just make a PostprocessorNeumannBC then.

**Comment by aeslaughter** [comment]
This runs the full problem from Step 9, but instead of computing thermal conductivity from a Material it is computed by a sub-app.

The master app feeds a gradient and a temperature to the sub. The sub uses this to compute an effective conductivity using a ClosePackIC. The sub feeds back thermal conductivity.

The problem with this example is that the k_eff calculation doesn't really change all that much, so there are convergence "issues" (i.e., there is nothing to solve). I tried putting in k_water as a function of temperature, but that didn't help much.

**Comment by aeslaughter** [comment]
Just a thought for discussion, should tutorials be in a different repository? We could make it a submodule in MOOSE, I just added 2700 lines that is almost all copy and paste. 

And/or could we use symlinks or something on some of these files so that we don't have copies of copies.

**Comment by aeslaughter** [comment]
What if at the micro-structure level we artificially grow the ball size with time? It is not physical, but it would fairly easy to implement.

**Comment by aeslaughter** [comment]
I am doing some work on the micro-calculation to add noise, I will re-open.

**Comment by aeslaughter** [comment]
This is much better now, I created a `RandomCorrosion` aux kernel that removes steel at random, and it does so more as a function of temperature.

I will post a picture/video soonish.

File: framework/src/bcs/NeumannBC.C (hunk 1 above)
Comment (permcody) [inline framework/src/bcs/NeumannBC.C]: Looks like you didn't undo the changes to Neumann yet.

File: framework/include/bcs/NeumannBC.h [hunk 3]
```diff
@@ -41,7 +41,7 @@ class NeumannBC : public IntegratedBC
   virtual Real computeQpResidual();
 
   /// Value of grad(u) on the boundary.
-  const Real & _value;
+  const PostprocessorValue & _value;
```
Comment (jwpeterson) [inline framework/include/bcs/NeumannBC.h]: I think @aeslaughter is still going to undo this line...

**Comment by jwpeterson** [comment]
I'm still not super happy about the way we have 10 almost-but-not-quite identical copies of each tutorial example... but I don't have any better solution.

File: framework/include/bcs/NeumannBC.h (hunk 3 above)
Comment (aeslaughter) [inline framework/include/bcs/NeumannBC.h]: I will, thought I did.

**Comment by aeslaughter** [comment]
I created a MultiApp step, it probably could be improved, but I feel it is worthy of being added.

I also created symlinks to files that should not change between the steps; hopefully, this will make it a bit easier to maintain.

File: test/tests/postprocessors/real_parameter_reporter/real_parameter_reporter.i~HEAD [hunk 4]
```diff
@@ -0,0 +1,63 @@
+[Mesh]
```
Comment (permcody) [inline test/tests/postprocessors/real_parameter_reporter/real_parameter_reporter.i~HEAD]: Are you committing temporary files?

**Comment by jwpeterson** [comment]
Just to be clear, we are voting on this question:

> I created a MultiApp step, it probably could be improved, but I feel it is worthy of being added.

**Comment by permcody** [comment]
I vote yes as well. I suppose we can do a janitor checkin to clean up any temporaries.

**Comment by aeslaughter** [comment]
I just removed the temporary file, @permcody you passed my test.
